pkg/tso/keyspace_group_manager.go

```diff
@@ -881,3 +894,172 @@ func (kgm *KeyspaceGroupManager) finishSplitKeyspaceGroup(id uint32) error {
 	kgm.kgs[id] = splitGroup
 	return nil
 }
+
+func (kgm *KeyspaceGroupManager) finishMergeKeyspaceGroup(id uint32) error {
```

Comment:
By "the keyspace meta's config map still has it's group data which will block this keyspace to be split again in PatrolKeyspaceAssignment", I meant in the following code, if the keyspace already has a group ID, we simply skip the only place during split to set the group in the existing code base (w/o #6578), which has the problem if splitTargetID in re-split is different from the existing group of the keyspace. I agree that "which will block this keyspace to be split again" in my statement isn't accurate.

#6578 can fix the inconsistency between keyspace meta and keyspace group's keyspace list, but it also has problem, because SaveKeyspaceMeta() needs to be called in batches too, otherwise the len(keyspaces) in a split needs to be <= maxTxnOps which isn't practical even in dev with tens thousands of keyspaces. Adding batching in another place increases complexity again. 

We need to simplify the split/merge logic and I have some thoughts about it. Let's discuss at at the first thing of Tuesday. @JmPotato  @rleungx @lhy1024 


```
func (manager *Manager) PatrolKeyspaceAssignment() error {
...
			for _, ks := range keyspaces {
				if ks == nil {
					continue
				}
				patrolledKeyspaceCount++
				manager.metaLock.Lock(ks.Id)
				if ks.Config == nil {
					ks.Config = make(map[string]string, 1)
				} else if _, ok := ks.Config[TSOKeyspaceGroupIDKey]; ok {
					// If the keyspace already has a group ID, skip it.
					manager.metaLock.Unlock(ks.Id)
					continue
				}
                                 ...
                                 ks.Config[TSOKeyspaceGroupIDKey] = toString(utils.DefaultKeyspaceGroupID)
				err = manager.store.SaveKeyspaceMeta(txn, ks)

```

Reply:
After syncing offline, we'll downgrade 'group' field in keyspace meta to a flag indicating if a keyspace is a legacy keyspace and if it needs to be put into default keyspace group. With this design change, #6578 isn't necessary.